midend/lib/Utils/DIPUtils.cpp

```diff
@@ -410,6 +412,182 @@ Value customTanVal(OpBuilder &builder, Location loc, Value angleVal) {
+  Value a0 = builder.create<arith::MulFOp>(loc, affineMatrix[4], D);
+  Value a4 = builder.create<arith::MulFOp>(loc, affineMatrix[0], D);
+  affineMatrix[0] = a0;
+  affineMatrix[1] = builder.create<arith::MulFOp>(loc, affineMatrix[1], negD);
+  affineMatrix[3] = builder.create<arith::MulFOp>(loc, affineMatrix[3], negD);
+  affineMatrix[4] = a4;
+  Value m0pm2 =
+      builder.create<arith::MulFOp>(loc, affineMatrix[0], affineMatrix[2]);
+  Value m1pm5 =
+      builder.create<arith::MulFOp>(loc, affineMatrix[1], affineMatrix[5]);
+  Value negB1 = builder.create<arith::AddFOp>(loc, m0pm2, m1pm5);
+  Value m2pm3 =
+      builder.create<arith::MulFOp>(loc, affineMatrix[2], affineMatrix[3]);
+  Value m4pm5 =
+      builder.create<arith::MulFOp>(loc, affineMatrix[4], affineMatrix[5]);
+  Value negB2 = builder.create<arith::AddFOp>(loc, m2pm3, m4pm5);
+  affineMatrix[2] = builder.create<arith::NegFOp>(loc, negB1);
+  affineMatrix[5] = builder.create<arith::NegFOp>(loc, negB2);
+}
+
+// Controls affine transform application.
+void affineTransformController(OpBuilder &builder, Location loc,
+                               MLIRContext *ctx, Value input, Value output,
+                               SmallVector<Value, 6> affineMatrix,
+                               int64_t stride) {
+  VectorType vectorTyF32 = VectorType::get({stride}, FloatType::getF32(ctx));
+  VectorType vectorTyI32 = VectorType::get({stride}, IntegerType::get(ctx, 32));
+
+  Value c0Index = builder.create<arith::ConstantIndexOp>(loc, 0);
+  Value c1Index = builder.create<arith::ConstantIndexOp>(loc, 1);
+
+  inverseAffineMatrix(builder, loc, affineMatrix);
+
+  Value m0Vec =
+      builder.create<vector::SplatOp>(loc, vectorTyF32, affineMatrix[0]);
+  Value m2Vec =
+      builder.create<vector::SplatOp>(loc, vectorTyF32, affineMatrix[2]);
+  Value m3Vec =
+      builder.create<vector::SplatOp>(loc, vectorTyF32, affineMatrix[3]);
+  Value m5Vec =
+      builder.create<vector::SplatOp>(loc, vectorTyF32, affineMatrix[5]);
+
+  Value outputRow = builder.create<memref::DimOp>(loc, output, c0Index);
+  Value outputCol = builder.create<memref::DimOp>(loc, output, c1Index);
+
+  Value strideVal = builder.create<arith::ConstantIndexOp>(loc, stride);
+  Value outputColStrideRatio =
+      builder.create<arith::DivUIOp>(loc, outputCol, strideVal);
+  Value outputColMultiple = builder.create<arith::MulIOp>(
+      loc, builder.create<arith::AddIOp>(loc, outputColStrideRatio, c1Index),
+      strideVal);
+
+  Value xVecInitial = iotaVec0F32(builder, loc, stride);
+
+  MemRefType dynamicTypeI32 =
+      MemRefType::get(ShapedType::kDynamic, IntegerType::get(ctx, 32));
+
+  // compute x*m0+m2 and x*m3+m5 and store the results into xMm0 and xMm3
+  Value xMm0 =
+      builder.create<memref::AllocOp>(loc, dynamicTypeI32, outputColMultiple);
+  Value xMm3 =
+      builder.create<memref::AllocOp>(loc, dynamicTypeI32, outputColMultiple);
+
+  const int RSV_BITS = 5;
```

Comment:
There is a missing explanation of the RSV_BITS variable and why it is assigned a value of 5. Please add some comments here.